tests/languages/download_test.py

```diff
@@ -0,0 +1,191 @@
+test.bat""",
+]
+
+deps_invalid_length = [
+    """linux/amd64
+sha256-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+http://127.0.0.1:5555
+test.bat""",
+    """windows/amd64
+sha256-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+http://127.0.0.1:5555
+test.bat""",
+]
+
+deps_wrong_checksum = [
+    """linux/amd64
+sha256-oRJkj6Cr8nWIivZ9d4W+rVZt/aSW1l9YtxSVh+GtIHM=
+http://127.0.0.1:5555
+test.bat""",
+    """windows/amd64
+sha256-L63Nefq+fKVIm25IKqlcqbJmc1rrJD3dKhIvutFK+IA=
+http://127.0.0.1:5555
+test.bat""",
+]
+
+
+class HTTPRequestHandler(BaseHTTPRequestHandler):
+    def do_GET(self):
+        self.send_response(200)
+        self.end_headers()
+        if platform.system() == 'Windows':
+            self.wfile.write(batch_script)
+        else:
+            self.wfile.write(shellscript)
+
+
+class server():
+    def __init__(self):
+        self.httpd = HTTPServer(('localhost', 5555), HTTPRequestHandler)
+
+    def start(self):
+        thread.start_new_thread(self.httpd.serve_forever, ())
+
+    def stop(self):
+        self.httpd.shutdown()
+
+
+@pytest.fixture
+def download_dir(tmpdir):
+    with tmpdir.as_cwd():
+        prefix = tmpdir.join('prefix').ensure_dir()
+        prefix = Prefix(str(prefix))
+        yield prefix, tmpdir
+
+
+def test_download_unhealthy_download(download_dir: Iterator[Prefix]) -> None:
+    """
+    This check against various errors during download.
+    """
+    prefix, _ = download_dir
+    http = server()
+    http.start()
+
+    # This checks against invalid SRI string.
+    try:
+        download.install_environment(prefix, C.DEFAULT, deps_invalid_length)
+    except ValueError as err:
+        assert err.args[0] == 'Invalid checksum string length of 33 \
+for sha256, expected 32'
+
+    # This checks against mismatched checksum
+    try:
+        download.install_environment(prefix, C.DEFAULT, deps_wrong_checksum)
+    except ChecksumMismatchError as err:
+        assert err.actual == correct_checksum
+    http.stop()
+
+
+def test_download_healthy_health_check(download_dir: Iterator[Prefix]) -> None:
+    """
+    This performs a normal download and on-disk health check.
+    """
+    prefix, _ = download_dir
+    http = server()
+    http.start()
+    download.install_environment(prefix, C.DEFAULT, healthy_dependencies)
+    http.stop()
+    assert download.health_check(prefix, C.DEFAULT) is None
+
+
+def test_download_unhealthy_health_check(
+    download_dir: Iterator[Prefix],
+) -> None:
+    """
+    This is for simulating changes to pre-commit cache,
+    e.g. unexpected changes to the on-disk cache.
+    """
+    prefix, _ = download_dir
+    envdir = Path(
+        lang_base.environment_dir(
+            prefix, ENVIRONMENT_DIR,
+            get_default_version(),
+        ),
+    )
+    script = envdir / 'test.bat'
+    http = server()
+    http.start()
+    download.install_environment(prefix, C.DEFAULT, healthy_dependencies)
+    chmod(script, stat.S_IRUSR | stat.S_IWUSR)
```

Comment:
This should be in the `install_environment`, otherwise the `download` language is unusable as no downloaded files will be executable.

Reply:
This is giving read(stat.S_IRUSR) and **write**(stat.S_IWUSR) permission to the downloaded file, so I can simulate corrupting the file. 

By default I only give the downloaded file read and execute permission. See here: https://github.com/fangfufu/pre-commit/blob/81aaf89a80c2ad764a186b8479bdde48a580da02/pre_commit/languages/download.py#L253